### PR TITLE
feat!: use js-tokens instead of acorn

### PR DESCRIPTION
**Pull request by sheremet-va**
### Description

This PR replaces `acorn` with `js-tokens`. All tests are passing (I did update Vitest that changed snapshots, but I did it before running tests, so this is legitimate). I am opening this PR before I am fully ready (I didn't try updating `createIsLiteralPositionAcorn` yet) to also discuss some potential API changes:

1. `js-tokens` never throws an error, so do we need still the regexp version?
2. Tokens don't have any information about start/end - is it needed somewhere?
3. Do we maybe want to have a separate entry point like `strip-iteral/acorn` with the previous implementation, and specify acorn in peerDependencies?
4. I am also open to any review

### Linked Issues

- https://github.com/antfu/strip-literal/issues/5

### CODE REVIEW

**Comment by sheremet-va** [comment]
@antfu what do you think?

File: test/custom-fill.test.ts [hunk 1]
```diff
@@ -22,33 +22,7 @@ const c = \`aaaa\${foo}dddd\${bar}\`
                  
       const a = 'aaaa'
                     
-      const b = \\"----\\"
-
-      const c = \`aaaa\${foo}----\${bar}\`"
-    `)
-  })
-
-  it('regex', () => {
```
Comment (sheremet-va) [inline test/custom-fill.test.ts:29]: Removed because it never calls regexp

File: test/fixtures/forgiving.output.vue [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// mode: regex
```
Comment (sheremet-va) [inline test/fixtures/forgiving.output.vue:1]: I was surprised vue files worked 😄  (although tokens there are complete mess)

File: src/js-tokens.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
+import jsTokens, { type Token as JSToken } from 'js-tokens'
+import type { StripLiteralOptions } from './types'
+
+export function _stripLiteralJsTokens(code: string, options?: StripLiteralOptions) {
+  const FILL = options?.fillChar ?? ' '
+  const FILL_COMMENT = ' '
+  let result = ''
+
+  const filter = options?.filter ?? (() => true)
+
```
Comment (sheremet-va) [inline src/js-tokens.ts:11]: Acorn implementation has `fillupTo`, but I am not sure how it is used to be honest 🤔

File: src/js-tokens.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      token => token,
+      body => body.replace(/[^\n]/g, FILL_COMMENT),
+      true,
+    ],
+    RegularExpressionLiteral: [
+      token => token,
+      body => body.replace(/\/(.*)\/(\w?)$/g, (_, $1, $2) => `/${FILL.repeat($1.length)}/${$2}`),
+    ],
+    // `start${
+    TemplateHead: [
```
Comment (sheremet-va) [inline src/js-tokens.ts]: Template strings can also have new lines 🤔

**Comment by antfu** [comment]
Looks solid! It seems that regex mode is never used now. Could we find a counter-example, or maybe we could just remove the regex mode and always use js-tokens?

File: src/js-tokens.ts (hunk 3 above)
Comment (antfu) [reply src/js-tokens.ts:11]: In acorn AST the tokens could be discrete, we use that to fill with the original code between the gaps of two tokens. If js-tokens does not need it then no worries on it

**Comment by antfu** [comment]
For `createIsLiteralPositionAcorn`, I searched the entire GH https://github.com/search?q=createIsLiteralPositionAcorn&type=code and it seems no one is using it. I kind of forgot why I introduced it at the beginning, but I am ok with a big breaking change on the internal change. The expected behavior is not affected anyway :)

For `strip-iteral/acorn`, I think it's fine to remove it. People who absolutely need it can use the old version

**Comment by antfu** [comment]
Did the bench:
image

To my surprise the regex is the slowest. I am ok with js-tokens being a bit slower than acorn for the better error tolerance (otherwise we are fallback to regex often)

**Comment by sheremet-va** [comment]
Do you have this benchmark somewhere? I can try optimizing the new implementation a bit more (using `if` instead of factory might be faster, but I don't know how fast)

**Comment by sheremet-va** [comment]
> Do you have this benchmark somewhere? I can try optimizing the new implementation a bit more (using if instead of factory might be faster, but I don't know how fast)

I see now you already pushed it :D

> Could we find a counter-example, or maybe we could just remove the regex mode and always use js-tokens?

As I said in the issue description, it never throws an error. I am fine with removing the regexp implementation. 

But does anyone use `tokens` from the detailed version?

**Comment by sheremet-va** [comment]
@antfu acorn and regexp are removed now, ready for a review. I removed factory with restructuring, so it's a _little_ bit faster now

**Review by antfu — APPROVED** [review]

**Pull request merged**